### PR TITLE
Print more information for TLS errors

### PR DESCRIPTION
**Pull request by ljrk0**
The aatxe/irc crate defines a crate-wide Error enum in irc/src/error.rs, wrapping different errors from various kinds of sources, such as TLS errors by tls-native.  When `.to_string()` is invoked on such an enum, the error message using `#[error("...")]` is displayed.  However, they wrap the actual Tls error enum inside with its own `.to_string()` function.  In case of such errors, also include this information in our error message which is much more helpful for debugging.

image

Note that perhaps it'd be better to fix this upstream, it seems that the macro can at least also access fields from the enum, as seen here: https://github.com/aatxe/irc/blob/develop/src/error.rs#L76. Maybe something like

```
#[error("a TLS error occured: {}", tlserror)
```

with a corresponding field in the enum upstream would work, but my Rust foo is quite limited.

### CODE REVIEW

**Comment by ljrk0** [comment]
I've created an upstream PR as the code was much more easy than anticipated :)

**Comment by tarkah** [comment]
Thanks for opening the PR!

It's idiomatic to not expose all error context in the `Display` implementation at each level. It's up to the consumer to decide if they want to dig into this context more, and if so, we can easily pattern match for the `Error::Tls` variant and print it's inner error. 

I'd prefer to see the original approach vs having the maintainer pollute the top-level error display with all context, which isn't suitable for all consumers.

So without changing upstream, a good solution might be:

```rs
let error = match e {
  irc::Error::Tls(e) => format!("a TLS error occurred: {e}"),
  _ => e.to_string(),
};
```

**Comment by ljrk0** [comment]
Thank you for your comment & time!

Ah, I see – I was just mirroring what other errors in the irc crate do, but I'm a noob when it comes to app design :)  From what I can understand, you propose having the "complete" error in the log and the "top level" error in the UI?  That would definitely work for me, especially given those TLS errors often even being multiline.

**Comment by tarkah** [comment]
> Thank you for your comment & time!
> 
> Ah, I see – I was just mirroring what other errors in the irc crate do, but I'm a noob when it comes to app design :)  From what I can understand, you propose having the "complete" error in the log and the "top level" error in the UI?  That would definitely work for me, especially given those TLS errors often even being multiline.

No problem! I think it's fine to use that new error string in both places. I think it'd be helpful to see it in the UI buffer since it's probably a common error where the context might help.

The match based example I provided would be idiomatic and work well.

File: data/src/stream.rs [hunk 1]
```diff
@@ -98,12 +98,19 @@ pub async fn run(server: server::Entry, mut sender: mpsc::Sender<Update>) -> Nev
                         };
                     }
                     Err(e) => {
+                        let e = e.to_string() + &match e {
+                            // unwrap Tls-specific error enums to access more error info
+                            irc::error::Error::Tls(ne) => {
+                                ": ".to_string() + &ne.to_string()
+                            },
+                            _ => "".to_string()
+                        };
```
Comment (tarkah) [inline data/src/stream.rs:107]: You can do the following which will reduce the # of allocations and is a little easier to read:

```rs
let error = match e {
  irc::Error::Tls(e) => format!("a TLS error occurred: {e}"),
  _ => e.to_string(),
};
```

**Review by tarkah — CHANGES_REQUESTED** [review]

**Comment by tarkah** [comment]
@ljrk0 Looks great, thanks for the contribution :tada:

**Review by tarkah — APPROVED** [review]

**Pull request merged**